### ISSUE TITLE
Combining the results from different runs

### ISSUE THREAD

**Issue by mooahmed** [automl/auto-sklearn#635]
Wonderful package.
My goal is to use it in parallel on multiple machines where they share the same storage.
To verify that the results will be reused from the different runs,  I am running a small experiment as follows:
I run autosklearn sequentially with seed=1 and get model1 then seed=2 to get model2 and seed=3 to get model3.
If share_model=True, then I expect: 
a) the results of model1 to be a subset of the results of model2 
b) the results of model1 to be a subset of the results of model3.
c) the results of model2 to be a subset of the results of model3.

When I run this code, conditions "a" and "b" are satisfied but "c" is not.

This is the code that I used:
```

total_time = 50
time_per_task = 25
memory_limit= 32000  #does not work on jupyter lab for smaller values?

X, y = sklearn.datasets.load_digits(return_X_y=True)
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, random_state=123)

tmp_folder='./tmp1'
output_folder='./out1' 
                                                          
seed = 1

automl1 = autosklearn.classification.AutoSklearnClassifier( time_left_for_this_task=total_time,
                                                          ml_memory_limit= memory_limit,  
                                                          per_run_time_limit=time_per_task,
                                                          tmp_folder=tmp_folder, 
                                                          output_folder=output_folder, 
                                                          seed=seed, 
                                                          shared_mode=True, 
                                                          ensemble_size=0,
                                                          initial_configurations_via_metalearning=0
                                                         )
automl1.fit(X_train, y_train)

seed = 2
automl2 = autosklearn.classification.AutoSklearnClassifier( time_left_for_this_task=total_time,
                                                          ml_memory_limit= memory_limit,
                                                          per_run_time_limit=time_per_task,
                                                          tmp_folder=tmp_folder, 
                                                          output_folder=output_folder, 
                                                          seed=seed, 
                                                          shared_mode=True, 
                                                          ensemble_size=0,
                                                          initial_configurations_via_metalearning=0
                                                         )
automl2.fit(X_train, y_train)

seed = 3
automl3 = autosklearn.classification.AutoSklearnClassifier( time_left_for_this_task=total_time,
                                                          ml_memory_limit= memory_limit,
                                                          per_run_time_limit=time_per_task,
                                                          tmp_folder=tmp_folder, 
                                                          output_folder=output_folder, 
                                                          seed=seed, 
                                                          shared_mode=True, 
                                                          ensemble_size=0,
                                                          initial_configurations_via_metalearning=0
                                                         )
automl3.fit(X_train, y_train)

# Combine the results from the 3 runs
seed = 4
ensemble = autosklearn.classification.AutoSklearnClassifier( time_left_for_this_task=total_time,
                                                          ml_memory_limit= memory_limit,
                                                          per_run_time_limit=time_per_task,
                                                          tmp_folder=tmp_folder, 
                                                          output_folder=output_folder, 
                                                          seed=seed, 
                                                          shared_mode=True, 
                                                          ensemble_size=1,
                                                          initial_configurations_via_metalearning=0
                                                         )

ensemble.fit(X_train, y_train)


print("Number of evaluated configurations")
print("for model1 = ",len(automl1.cv_results_['mean_test_score']))
print("for model2 = ",len(automl2.cv_results_['mean_test_score']))
print("for model3 = ",len(automl3.cv_results_['mean_test_score']))
print("for ensemble = ",len(ensemble.cv_results_['mean_test_score']))

print("Number of common configurations")
print("model 1 and model 2 =",len([x for x in automl1.cv_results_['mean_test_score'] if x in automl2.cv_results_['mean_test_score']]))
print("model 1 and model 3 =",len([x for x in automl1.cv_results_['mean_test_score'] if x in automl3.cv_results_['mean_test_score']]))
print("model 2 and model 3 =",len([x for x in automl2.cv_results_['mean_test_score'] if x in automl3.cv_results_['mean_test_score']]))

```
And  this is  the output:

```
Number of evaluated configurations
for model1 =  12
for model2 =  17
for model3 =  20
for ensemble =  19

Number of common configurations
model 1 and model 2 = 12
model 1 and model 3 = 12
model 2 and model 3 = 15

```

Another question, why is the mean_fit time is different for model1 and model2. It is close but different.
I thought all the results of model1 are reused as is for model2.

```
print("Model mean fit times")
print("model1 = ",automl1.cv_results_['mean_fit_time'])
print("model2 = ",automl2.cv_results_['mean_fit_time'])
```
In the output why is the first value differnt between models 1 and 2? :
```
model1 =  [ 0.8074  6.3149  2.5494  1.2129  1.7692  0.9731  0.8203  6.61    1.0786
  5.4731  0.8178  1.0049]
model2 =  [  0.819    6.3149   2.5494   1.2129   1.7692   0.9731   0.8203   6.61
   1.0786   5.4731   0.8178   1.0049   3.1685  15.5866   1.1651   0.7671
  11.0149]
```
A more weird behavior for the different runs is if I change the random seed1=10.
All the results from the different runs are totally different.

Maybe, I am missing something.
I appreciate your help.

**Comment by mfeurer** [comment]
Excellent questions. I don't have an answer to these right now, but we'll have a look at this.

**Comment by gui-miotto** [comment]
Dear @mooahmed , I will start working on this issue now. But before I do that, since you open it six months ago, I just wanted to ask you if you have gained any more insights into this problem since then. Have you noticed any other strange/interesting behavior related to this issue?

**Comment by gui-miotto** [comment]
@mooahmed , thanks for the easy to understand example and for bringing up these issues.
First, about the set/subset of configurations, I tried to reproduce this weird behavior, but couldn't. I tested many seeds and always got the expected answer:
```
Number of evaluated configurations
for model1 =  X
for model2 =  Y
for model3 =  Z

Number of common configurations
model 1 and model 2 = X
model 1 and model 3 = X
model 2 and model 3 = Y
```
I ran your example with auto-sklearn's latest version of the development branch, which is pretty much the next release that should be out in a couple of days. So what I recommend you to do is to updated your auto-sklearn (as soon as the new version sees the light of day) and rerun your code. If the issue persists, we would be glad to take a look at it again.

About the `mean_fit_time`, this is a SMAC thing. You can avoid this problem by specifying `smac_scenario_args = {'initial_incumbent': 'RANDOM'}` in the constructor of `AutoSklearnClassifier`.
 
To sum up, here is the code I ran and the output I got:
```python
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from autosklearn.classification import AutoSklearnClassifier

total_time = 50
time_per_task = 25
memory_limit= 32000  #does not work on jupyter lab for smaller values?

X, y = load_digits(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=123)

tmp_folder='./tmp1'
output_folder='./out1' 
                                                          
seed = 1

automl1 = AutoSklearnClassifier( 
    time_left_for_this_task=total_time,
    ml_memory_limit= memory_limit,  
    per_run_time_limit=time_per_task,
    tmp_folder=tmp_folder, 
    output_folder=output_folder, 
    seed=seed, 
    shared_mode=True, 
    ensemble_size=0,
    initial_configurations_via_metalearning=0,
    )
automl1.fit(X_train, y_train)

seed = 2
automl2 = AutoSklearnClassifier( 
    time_left_for_this_task=total_time,
    ml_memory_limit= memory_limit,
    per_run_time_limit=time_per_task,
    tmp_folder=tmp_folder, 
    output_folder=output_folder, 
    seed=seed, 
    shared_mode=True, 
    ensemble_size=0,
    initial_configurations_via_metalearning=0,
    smac_scenario_args = {'initial_incumbent': 'RANDOM'}  # Added argument
    )
automl2.fit(X_train, y_train)

seed = 3
automl3 = AutoSklearnClassifier( 
    time_left_for_this_task=total_time,
    ml_memory_limit= memory_limit,
    per_run_time_limit=time_per_task,
    tmp_folder=tmp_folder, 
    output_folder=output_folder, 
    seed=seed, 
    shared_mode=True, 
    ensemble_size=0,
    initial_configurations_via_metalearning=0,
    smac_scenario_args = {'initial_incumbent': 'RANDOM'}  # Added argument
    )
automl3.fit(X_train, y_train)

# Combine the results from the 3 runs
seed = 4
ensemble = AutoSklearnClassifier( 
    time_left_for_this_task=total_time,
    ml_memory_limit= memory_limit,
    per_run_time_limit=time_per_task,
    tmp_folder=tmp_folder, 
    output_folder=output_folder, 
    seed=seed, 
    shared_mode=True, 
    ensemble_size=1,
    initial_configurations_via_metalearning=0
    )

ensemble.fit(X_train, y_train)


print("Number of evaluated configurations")
print("for model1 = ",len(automl1.cv_results_['mean_test_score']))
print("for model2 = ",len(automl2.cv_results_['mean_test_score']))
print("for model3 = ",len(automl3.cv_results_['mean_test_score']))
print("for ensemble = ",len(ensemble.cv_results_['mean_test_score']))

print("Number of common configurations")
print("model 1 and model 2 =",len([x for x in automl1.cv_results_['mean_test_score'] if x in automl2.cv_results_['mean_test_score']]))
print("model 1 and model 3 =",len([x for x in automl1.cv_results_['mean_test_score'] if x in automl3.cv_results_['mean_test_score']]))
print("model 2 and model 3 =",len([x for x in automl2.cv_results_['mean_test_score'] if x in automl3.cv_results_['mean_test_score']]))

print("Model mean fit times")
print("model1 = ",automl1.cv_results_['mean_fit_time'])
print("model2 = ",automl2.cv_results_['mean_fit_time'])
```

OUTPUT:
Number of evaluated configurations
for model1 =  6
for model2 =  13
for model3 =  19
for ensemble =  22
Number of common configurations
model 1 and model 2 = 6
model 1 and model 3 = 6
model 2 and model 3 = 13
Model mean fit times
model1 =  [ **0.85856581  1.08830547  3.45141554  0.93370342 25.0317173   1.94852448**]
model2 =  [ 0.65213227  **0.85856581  1.08830547  3.45141554  0.93370342 25.0317173
  1.94852448**  0.64790392 25.02988791  1.08440471  1.4163847   0.62608266
  1.77576828]

**Comment by gui-miotto** [comment]
@mfeurer , should we close this one?

**Comment by mooahmed** [comment]
@gui-miotto 
Thanks a lot for taking the time to check that. I will check it again with the latest released autosklearn version.
Regarding the mean fit time, what does `smac_scenario_args = {'initial_incumbent': 'RANDOM'}` do exactly?

**Comment by mfeurer** [comment]
SMAC first runs the default configuration and the proceeds to the meta-learning configurations. Only then it reads the results from other invocations of SMAC/Auto-sklearn. This means that the first and second invocation of Auto-sklearn both execute the same configuration (the default). Therefore, the first run had two different runtimes. 

If you pass the flag `'initial_incumbent': 'RANDOM' SMAC will not use a default configuration to start, but a random configuration. That's why in @gui-miotto example the `model2` had a vastly different runtime.

In any case, we figured that the first runtime will always differ in the current setting.

**Comment by mfeurer** [comment]
Closing this as it seems to be fixed. Please reopen if you have any further questions.
